### ISSUE TITLE
[libffi] Build error on x64-windows

### ISSUE THREAD

**Issue by henrygab** [microsoft/vcpkg#35966]
This is a clean install/setup of vcpkg.  Any tips to get unblocked would be appreciated.

### Steps to repro
1. VS2022 / Win11 x64
2. In VS2022, create a new cmake windows project at `C:\src\bug`
3. In developer console (within VS2022) at `c:\src\bug`:
   ```
   git clone https://github.com/microsoft/vcpkg
   .\vcpkg\bootstrap-vcpkg.bat
   .\vcpkg\vcpkg.exe update
   .\vcpkg\vcpkg.exe install boost
   ```

## Issue_Body.md contents

Package: libffi:x64-windows@3.4.4#6

**Host Environment**

- Host: x64-windows
- Compiler: MSVC 19.38.33133.0
-    vcpkg-tool version: 2023-12-12-1c9ec1978a6b0c2b39c9e9554a96e3e275f7556e
    vcpkg-scripts version: 839722725 2023-12-29 (3 days ago)

**To Reproduce**

`vcpkg install boost`
**Failure logs**

```
-- Downloading https://github.com/libffi/libffi/releases/download/v3.4.4/libffi-3.4.4.tar.gz -> libffi-3.4.4.tar.gz...
-- Extracting source C:/src/bug/vcpkg/downloads/libffi-3.4.4.tar.gz
-- Applying patch dll-bindir.diff
-- Applying patch fix_undefind_func.patch
-- Using source at C:/src/bug/vcpkg/buildtrees/libffi/src/libffi-3-b5bf0e8df1.clean
-- Found external ninja('1.11.0').
-- Getting CMake variables for x64-windows
-- Getting CMake variables for x64-windows
-- Downloading https://mirror.msys2.org/msys/x86_64/autoconf-wrapper-20221207-1-any.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/autoconf-wrapper-20221207-1-any.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/autoconf-wrapper-20221207-1-any.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/autoconf-wrapper-20221207-1-any.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/autoconf-wrapper-20221207-1-any.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/autoconf-wrapper-20221207-1-any.pkg.tar.zst -> msys2-autoconf-wrapper-20221207-1-any.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/automake-wrapper-20221207-1-any.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/automake-wrapper-20221207-1-any.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/automake-wrapper-20221207-1-any.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/automake-wrapper-20221207-1-any.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/automake-wrapper-20221207-1-any.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/automake-wrapper-20221207-1-any.pkg.tar.zst -> msys2-automake-wrapper-20221207-1-any.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/binutils-2.41-3-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/binutils-2.41-3-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/binutils-2.41-3-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/binutils-2.41-3-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/binutils-2.41-3-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/binutils-2.41-3-x86_64.pkg.tar.zst -> msys2-binutils-2.41-3-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/libtool-2.4.7-3-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/libtool-2.4.7-3-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/libtool-2.4.7-3-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/libtool-2.4.7-3-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/libtool-2.4.7-3-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/libtool-2.4.7-3-x86_64.pkg.tar.zst -> msys2-libtool-2.4.7-3-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/make-4.4.1-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/make-4.4.1-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/make-4.4.1-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/make-4.4.1-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/make-4.4.1-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/make-4.4.1-1-x86_64.pkg.tar.zst -> msys2-make-4.4.1-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/pkgconf-2.1.0-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/pkgconf-2.1.0-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/pkgconf-2.1.0-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/pkgconf-2.1.0-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/pkgconf-2.1.0-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/pkgconf-2.1.0-1-x86_64.pkg.tar.zst -> msys2-pkgconf-2.1.0-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/which-2.21-4-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/which-2.21-4-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/which-2.21-4-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/which-2.21-4-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/which-2.21-4-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/which-2.21-4-x86_64.pkg.tar.zst -> msys2-which-2.21-4-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/bash-5.2.021-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/bash-5.2.021-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/bash-5.2.021-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/bash-5.2.021-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/bash-5.2.021-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/bash-5.2.021-1-x86_64.pkg.tar.zst -> msys2-bash-5.2.021-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/coreutils-8.32-5-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/coreutils-8.32-5-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/coreutils-8.32-5-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/coreutils-8.32-5-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/coreutils-8.32-5-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/coreutils-8.32-5-x86_64.pkg.tar.zst -> msys2-coreutils-8.32-5-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/file-5.45-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/file-5.45-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/file-5.45-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/file-5.45-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/file-5.45-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/file-5.45-1-x86_64.pkg.tar.zst -> msys2-file-5.45-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/gawk-5.3.0-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/gawk-5.3.0-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/gawk-5.3.0-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/gawk-5.3.0-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/gawk-5.3.0-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/gawk-5.3.0-1-x86_64.pkg.tar.zst -> msys2-gawk-5.3.0-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/grep-1~3.0-6-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/grep-1~3.0-6-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/grep-1~3.0-6-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/grep-1~3.0-6-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/grep-1~3.0-6-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/grep-1~3.0-6-x86_64.pkg.tar.zst -> msys2-grep-1~3.0-6-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/gzip-1.13-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/gzip-1.13-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/gzip-1.13-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/gzip-1.13-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/gzip-1.13-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/gzip-1.13-1-x86_64.pkg.tar.zst -> msys2-gzip-1.13-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/diffutils-3.10-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/diffutils-3.10-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/diffutils-3.10-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/diffutils-3.10-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/diffutils-3.10-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/diffutils-3.10-1-x86_64.pkg.tar.zst -> msys2-diffutils-3.10-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/sed-4.9-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/sed-4.9-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/sed-4.9-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/sed-4.9-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/sed-4.9-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/sed-4.9-1-x86_64.pkg.tar.zst -> msys2-sed-4.9-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/msys2-runtime-3.4.9-3-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/msys2-runtime-3.4.9-3-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/msys2-runtime-3.4.9-3-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/msys2-runtime-3.4.9-3-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/msys2-runtime-3.4.9-3-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/msys2-runtime-3.4.9-3-x86_64.pkg.tar.zst -> msys2-msys2-runtime-3.4.9-3-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/autoconf2.71-2.71-3-any.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/autoconf2.71-2.71-3-any.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/autoconf2.71-2.71-3-any.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/autoconf2.71-2.71-3-any.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/autoconf2.71-2.71-3-any.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/autoconf2.71-2.71-3-any.pkg.tar.zst -> msys2-autoconf2.71-2.71-3-any.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/automake1.16-1.16.5-1-any.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/automake1.16-1.16.5-1-any.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/automake1.16-1.16.5-1-any.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/automake1.16-1.16.5-1-any.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/automake1.16-1.16.5-1-any.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/automake1.16-1.16.5-1-any.pkg.tar.zst -> msys2-automake1.16-1.16.5-1-any.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/libiconv-1.17-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/libiconv-1.17-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/libiconv-1.17-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/libiconv-1.17-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/libiconv-1.17-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/libiconv-1.17-1-x86_64.pkg.tar.zst -> msys2-libiconv-1.17-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/libintl-0.22.4-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/libintl-0.22.4-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/libintl-0.22.4-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/libintl-0.22.4-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/libintl-0.22.4-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/libintl-0.22.4-1-x86_64.pkg.tar.zst -> msys2-libintl-0.22.4-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/zlib-1.3-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/zlib-1.3-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/zlib-1.3-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/zlib-1.3-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/zlib-1.3-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/zlib-1.3-1-x86_64.pkg.tar.zst -> msys2-zlib-1.3-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/tar-1.35-2-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/tar-1.35-2-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/tar-1.35-2-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/tar-1.35-2-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/tar-1.35-2-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/tar-1.35-2-x86_64.pkg.tar.zst -> msys2-tar-1.35-2-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/findutils-4.9.0-3-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/findutils-4.9.0-3-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/findutils-4.9.0-3-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/findutils-4.9.0-3-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/findutils-4.9.0-3-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/findutils-4.9.0-3-x86_64.pkg.tar.zst -> msys2-findutils-4.9.0-3-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/gmp-6.3.0-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/gmp-6.3.0-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/gmp-6.3.0-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/gmp-6.3.0-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/gmp-6.3.0-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/gmp-6.3.0-1-x86_64.pkg.tar.zst -> msys2-gmp-6.3.0-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/gcc-libs-13.2.0-2-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/gcc-libs-13.2.0-2-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/gcc-libs-13.2.0-2-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/gcc-libs-13.2.0-2-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/gcc-libs-13.2.0-2-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/gcc-libs-13.2.0-2-x86_64.pkg.tar.zst -> msys2-gcc-libs-13.2.0-2-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/libbz2-1.0.8-4-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/libbz2-1.0.8-4-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/libbz2-1.0.8-4-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/libbz2-1.0.8-4-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/libbz2-1.0.8-4-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/libbz2-1.0.8-4-x86_64.pkg.tar.zst -> msys2-libbz2-1.0.8-4-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/liblzma-5.4.5-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/liblzma-5.4.5-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/liblzma-5.4.5-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/liblzma-5.4.5-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/liblzma-5.4.5-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/liblzma-5.4.5-1-x86_64.pkg.tar.zst -> msys2-liblzma-5.4.5-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/libzstd-1.5.5-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/libzstd-1.5.5-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/libzstd-1.5.5-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/libzstd-1.5.5-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/libzstd-1.5.5-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/libzstd-1.5.5-1-x86_64.pkg.tar.zst -> msys2-libzstd-1.5.5-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/libreadline-8.2.007-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/libreadline-8.2.007-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/libreadline-8.2.007-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/libreadline-8.2.007-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/libreadline-8.2.007-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/libreadline-8.2.007-1-x86_64.pkg.tar.zst -> msys2-libreadline-8.2.007-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/mpfr-4.2.1-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/mpfr-4.2.1-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/mpfr-4.2.1-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/mpfr-4.2.1-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/mpfr-4.2.1-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/mpfr-4.2.1-1-x86_64.pkg.tar.zst -> msys2-mpfr-4.2.1-1-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/libpcre-8.45-4-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/libpcre-8.45-4-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/libpcre-8.45-4-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/libpcre-8.45-4-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/libpcre-8.45-4-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/libpcre-8.45-4-x86_64.pkg.tar.zst -> msys2-libpcre-8.45-4-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/m4-1.4.19-2-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/m4-1.4.19-2-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/m4-1.4.19-2-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/m4-1.4.19-2-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/m4-1.4.19-2-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/m4-1.4.19-2-x86_64.pkg.tar.zst -> msys2-m4-1.4.19-2-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/perl-5.36.1-2-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/perl-5.36.1-2-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/perl-5.36.1-2-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/perl-5.36.1-2-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/perl-5.36.1-2-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/perl-5.36.1-2-x86_64.pkg.tar.zst -> msys2-perl-5.36.1-2-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/ncurses-6.4-2-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/ncurses-6.4-2-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/ncurses-6.4-2-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/ncurses-6.4-2-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/ncurses-6.4-2-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/ncurses-6.4-2-x86_64.pkg.tar.zst -> msys2-ncurses-6.4-2-x86_64.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/libxcrypt-4.4.36-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/libxcrypt-4.4.36-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/libxcrypt-4.4.36-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/libxcrypt-4.4.36-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/libxcrypt-4.4.36-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/libxcrypt-4.4.36-1-x86_64.pkg.tar.zst -> msys2-libxcrypt-4.4.36-1-x86_64.pkg.tar.zst...
-- Applying patch C:/src/bug/vcpkg/scripts/cmake/compile_wrapper_consider_clang-cl.patch
-- Using msys root at C:/src/bug/vcpkg/downloads/tools/msys2/5f36c2d65b270362
-- Downloading https://mirror.msys2.org/mingw/mingw64/mingw-w64-x86_64-pkgconf-1~2.1.0-1-any.pkg.tar.zst;https://repo.msys2.org/mingw/mingw64/mingw-w64-x86_64-pkgconf-1~2.1.0-1-any.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/mingw/mingw64/mingw-w64-x86_64-pkgconf-1~2.1.0-1-any.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/mingw/mingw64/mingw-w64-x86_64-pkgconf-1~2.1.0-1-any.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/mingw/mingw64/mingw-w64-x86_64-pkgconf-1~2.1.0-1-any.pkg.tar.zst;https://mirror.selfnet.de/msys2/mingw/mingw64/mingw-w64-x86_64-pkgconf-1~2.1.0-1-any.pkg.tar.zst -> mingw-w64-x86_64-pkgconf-1~2.1.0-1-any.pkg.tar.zst...
-- Using cached msys2-msys2-runtime-3.4.9-3-x86_64.pkg.tar.zst.
-- Using msys root at C:/src/bug/vcpkg/downloads/tools/msys2/023cdb3ca06f77f2
-- Configuring x64-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: C:/src/bug/vcpkg/downloads/tools/msys2/5f36c2d65b270362/usr/bin/bash.exe --noprofile --norc --debug -c "V=1 CPP='compile cl.exe -E' CC='compile cl.exe' CC_FOR_BUILD='compile cl.exe' CPP_FOR_BUILD='compile cl.exe -E' CXX_FOR_BUILD='compile cl.exe' CXX='compile cl.exe' RC='windres-rc rc.exe' WINDRES='windres-rc rc.exe' AR='ar-lib lib.exe' LD='link.exe -verbose' RANLIB=':' STRIP=':' NM='dumpbin.exe -symbols -headers' DLLTOOL='link.exe -verbose -dll' CCAS=':' AS=':' ./../src/libffi-3-b5bf0e8df1.clean/configure --build=x86_64-pc-mingw32 \"--enable-portable-binary\" \"--disable-docs\" \"--disable-multi-os-directory\" \"CFLAGS=${CFLAGS} -DFFI_BUILDING -DFFI_BUILDING_DLL\" \"CCAS=msvcc.sh -m64\" \"ac_cv_prog_ac_ct_STRIP=:\" \"gl_cv_double_slash_root=yes\" \"ac_cv_func_memmove=yes\" \"--disable-silent-rules\" \"--verbose\" \"--enable-shared\" \"--disable-static\" \"--prefix=/C/src/bug/vcpkg/installed/x64-windows/debug\" \"--bindir=\\${prefix}/../tools/libffi/debug/bin\" \"--sbindir=\\${prefix}/../tools/libffi/debug/sbin\" \"--libdir=\\${prefix}/lib\" \"--includedir=\\${prefix}/../include\" \"--datarootdir=\\${prefix}/share/libffi\""
    Working Directory: C:/src/bug/vcpkg/buildtrees/libffi/x64-windows-dbg
    Error code: 127
    See logs for more information:

Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_make.cmake:832 (vcpkg_execute_required_process)
  ports/libffi/portfile.cmake:45 (vcpkg_configure_make)
  scripts/ports.cmake:170 (include)



```

</p>
</details>

**Comment by henrygab** [comment]
The following override allowed me to make forward progress, albeit without the latest version of `ffilib`.   I forced an old versoin of `ffilib` by adding the following `overrides` section to my `vcpkg.json`:

```json
    "dependencies": [
        "boost",
        "boost-multiprecision"
    ],
    "overrides": [
        {
            "name": "libffi",
            "version": "3.3#9"
        }
    ]
```

While this is an unsupported configuration, hopefully it helps to narrow the root cause.

**Comment by dg0yt** [comment]
The old version doesn't help to narrow the root cause.

> ~~~
> -- Configuring x64-windows-dbg
> CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
>     Command failed: C:/src/bug/vcpkg/downloads/tools/msys2/5f36c2d65b270362/usr/bin/bash.exe --noprofile --norc --debug -c "V=1 CPP='compile cl.exe -E' CC='compile cl.exe' CC_FOR_BUILD='compile cl.exe' CPP_FOR_BUILD='compile cl.exe -E' CXX_FOR_BUILD='compile cl.exe' CXX='compile cl.exe' RC='windres-rc rc.exe' WINDRES='windres-rc rc.exe' AR='ar-lib lib.exe' LD='link.exe -verbose' RANLIB=':' STRIP=':' NM='dumpbin.exe -symbols -headers' DLLTOOL='link.exe -verbose -dll' CCAS=':' AS=':' ./../src/libffi-3-b5bf0e8df1.clean/configure --build=x86_64-pc-mingw32 \"--enable-portable-binary\" \"--disable-docs\" \"--disable-multi-os-directory\" \"CFLAGS=${CFLAGS} -DFFI_BUILDING -DFFI_BUILDING_DLL\" \"CCAS=msvcc.sh -m64\" \"ac_cv_prog_ac_ct_STRIP=:\" \"gl_cv_double_slash_root=yes\" \"ac_cv_func_memmove=yes\" \"--disable-silent-rules\" \"--verbose\" \"--enable-shared\" \"--disable-static\" \"--prefix=/C/src/bug/vcpkg/installed/x64-windows/debug\" \"--bindir=\\${prefix}/../tools/libffi/debug/bin\" \"--sbindir=\\${prefix}/../tools/libffi/debug/sbin\" \"--libdir=\\${prefix}/lib\" \"--includedir=\\${prefix}/../include\" \"--datarootdir=\\${prefix}/share/libffi\""
>     Working Directory: C:/src/bug/vcpkg/buildtrees/libffi/x64-windows-dbg
>     Error code: 127
>     See logs for more information:
> 
> Call Stack (most recent call first):
>   scripts/cmake/vcpkg_configure_make.cmake:832 (vcpkg_execute_required_process)
>   ports/libffi/portfile.cmake:45 (vcpkg_configure_make)
>   scripts/ports.cmake:170 (include)

Note: No log files mentioned. vcpkg wasn't able to start `C:/src/bug/vcpkg/downloads/tools/msys2/5f36c2d65b270362/usr/bin/bash.exe`. Maybe some Antivirus or similar? Or interference by another msys2 runtime? This will probably also block other ports like `libiconv`, i.e. it may be unrelated to `libffi`.

**Comment by JonLiu1993** [comment]
@henrygab, Thanks for posting this issue, I can not repo your error on my latest vcpkg:
````
boost-yap:x64-windows package ABI: f44b96bdfe6375bb818ab2c4e6748d8a818f5dac0ae6b1526ced191ead7f9774
Installing 166/166 boost:x64-windows@1.83.0#1...
Building boost:x64-windows@1.83.0#1...
-- Performing post-build validation
Elapsed time to handle boost:x64-windows: 108 ms
boost:x64-windows package ABI:
Total install time: 33 min
PS F:\test\vcpkg> ./vcpkg --version
vcpkg package management program version 2023-12-12-1c9ec1978a6b0c2b39c9e9554a96e3e275f7556e

See LICENSE.txt for license information.
````
Check to see if you have anti-virus software turned on. If so, turn it off and try again.

**Comment by henrygab** [comment]
No 3rd party AV.  I went into defender settings and added an exclusion for the directory `C:\src\bug`.

`vcpkg install` still fails in the same manner.  See log (including `vcpkg.json`) by expanding the details

<details><summary>Issue_Body.md</summary>
<p>

Package: libffi:x64-windows@3.4.4#6

**Host Environment**

- Host: x64-windows
- Compiler: MSVC 19.38.33133.0
-    vcpkg-tool version: 2023-12-12-1c9ec1978a6b0c2b39c9e9554a96e3e275f7556e
    vcpkg-scripts version: 839722725 2023-12-29 (4 days ago)

**To Reproduce**

`vcpkg install `
**Failure logs**

```
-- Using cached libffi-3.4.4.tar.gz.
-- Cleaning sources at C:/src/bug/vcpkg/buildtrees/libffi/src/libffi-3-b5bf0e8df1.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/src/bug/vcpkg/downloads/libffi-3.4.4.tar.gz
-- Applying patch dll-bindir.diff
-- Applying patch fix_undefind_func.patch
-- Using source at C:/src/bug/vcpkg/buildtrees/libffi/src/libffi-3-b5bf0e8df1.clean
-- Found external ninja('1.11.0').
-- Getting CMake variables for x64-windows
-- Getting CMake variables for x64-windows
-- Using cached msys2-autoconf-wrapper-20221207-1-any.pkg.tar.zst.
-- Using cached msys2-automake-wrapper-20221207-1-any.pkg.tar.zst.
-- Using cached msys2-binutils-2.41-3-x86_64.pkg.tar.zst.
-- Using cached msys2-libtool-2.4.7-3-x86_64.pkg.tar.zst.
-- Using cached msys2-make-4.4.1-1-x86_64.pkg.tar.zst.
-- Using cached msys2-pkgconf-2.1.0-1-x86_64.pkg.tar.zst.
-- Using cached msys2-which-2.21-4-x86_64.pkg.tar.zst.
-- Using cached msys2-bash-5.2.021-1-x86_64.pkg.tar.zst.
-- Using cached msys2-coreutils-8.32-5-x86_64.pkg.tar.zst.
-- Using cached msys2-file-5.45-1-x86_64.pkg.tar.zst.
-- Using cached msys2-gawk-5.3.0-1-x86_64.pkg.tar.zst.
-- Using cached msys2-grep-1~3.0-6-x86_64.pkg.tar.zst.
-- Using cached msys2-gzip-1.13-1-x86_64.pkg.tar.zst.
-- Using cached msys2-diffutils-3.10-1-x86_64.pkg.tar.zst.
-- Using cached msys2-sed-4.9-1-x86_64.pkg.tar.zst.
-- Using cached msys2-msys2-runtime-3.4.9-3-x86_64.pkg.tar.zst.
-- Using cached msys2-autoconf2.71-2.71-3-any.pkg.tar.zst.
-- Using cached msys2-automake1.16-1.16.5-1-any.pkg.tar.zst.
-- Using cached msys2-libiconv-1.17-1-x86_64.pkg.tar.zst.
-- Using cached msys2-libintl-0.22.4-1-x86_64.pkg.tar.zst.
-- Using cached msys2-zlib-1.3-1-x86_64.pkg.tar.zst.
-- Using cached msys2-tar-1.35-2-x86_64.pkg.tar.zst.
-- Using cached msys2-findutils-4.9.0-3-x86_64.pkg.tar.zst.
-- Using cached msys2-gmp-6.3.0-1-x86_64.pkg.tar.zst.
-- Using cached msys2-gcc-libs-13.2.0-2-x86_64.pkg.tar.zst.
-- Using cached msys2-libbz2-1.0.8-4-x86_64.pkg.tar.zst.
-- Using cached msys2-liblzma-5.4.5-1-x86_64.pkg.tar.zst.
-- Using cached msys2-libzstd-1.5.5-1-x86_64.pkg.tar.zst.
-- Using cached msys2-libreadline-8.2.007-1-x86_64.pkg.tar.zst.
-- Using cached msys2-mpfr-4.2.1-1-x86_64.pkg.tar.zst.
-- Using cached msys2-libpcre-8.45-4-x86_64.pkg.tar.zst.
-- Using cached msys2-m4-1.4.19-2-x86_64.pkg.tar.zst.
-- Using cached msys2-perl-5.36.1-2-x86_64.pkg.tar.zst.
-- Using cached msys2-ncurses-6.4-2-x86_64.pkg.tar.zst.
-- Using cached msys2-libxcrypt-4.4.36-1-x86_64.pkg.tar.zst.
-- Using msys root at C:/src/bug/vcpkg/downloads/tools/msys2/5f36c2d65b270362
-- Using cached mingw-w64-x86_64-pkgconf-1~2.1.0-1-any.pkg.tar.zst.
-- Using cached msys2-msys2-runtime-3.4.9-3-x86_64.pkg.tar.zst.
-- Using msys root at C:/src/bug/vcpkg/downloads/tools/msys2/023cdb3ca06f77f2
-- Configuring x64-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: C:/src/bug/vcpkg/downloads/tools/msys2/5f36c2d65b270362/usr/bin/bash.exe --noprofile --norc --debug -c "V=1 CPP='compile cl.exe -E' CC='compile cl.exe' CC_FOR_BUILD='compile cl.exe' CPP_FOR_BUILD='compile cl.exe -E' CXX_FOR_BUILD='compile cl.exe' CXX='compile cl.exe' RC='windres-rc rc.exe' WINDRES='windres-rc rc.exe' AR='ar-lib lib.exe' LD='link.exe -verbose' RANLIB=':' STRIP=':' NM='dumpbin.exe -symbols -headers' DLLTOOL='link.exe -verbose -dll' CCAS=':' AS=':' ./../src/libffi-3-b5bf0e8df1.clean/configure --build=x86_64-pc-mingw32 \"--enable-portable-binary\" \"--disable-docs\" \"--disable-multi-os-directory\" \"CFLAGS=${CFLAGS} -DFFI_BUILDING -DFFI_BUILDING_DLL\" \"CCAS=msvcc.sh -m64\" \"ac_cv_prog_ac_ct_STRIP=:\" \"gl_cv_double_slash_root=yes\" \"ac_cv_func_memmove=yes\" \"--disable-silent-rules\" \"--verbose\" \"--enable-shared\" \"--disable-static\" \"--prefix=/C/src/bug/vcpkg_installed/x64-windows/debug\" \"--bindir=\\${prefix}/../tools/libffi/debug/bin\" \"--sbindir=\\${prefix}/../tools/libffi/debug/sbin\" \"--libdir=\\${prefix}/lib\" \"--includedir=\\${prefix}/../include\" \"--datarootdir=\\${prefix}/share/libffi\""
    Working Directory: C:/src/bug/vcpkg/buildtrees/libffi/x64-windows-dbg
    Error code: 127
    See logs for more information:

Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_make.cmake:832 (vcpkg_execute_required_process)
  C:/Users/lego/AppData/Local/vcpkg/registries/git-trees/982248d325407b09d26ad0d929f67fd0b873f984/portfile.cmake:45 (vcpkg_configure_make)
  scripts/ports.cmake:170 (include)

```
**Additional context**

<details><summary>vcpkg.json</summary>

```
{
  "version-string": "0.1",
  "dependencies": [
    "boost"
  ]
}

```
</details>


</p>
</details>

**Comment by henrygab** [comment]
FYI ... your logs do not appear to attempt a full repro (e.g., clean install Win11 + VS2022) ... or else the logs you showed were significantly truncated ... as I would expect the logs to show significant time taken downloading and installing many dependencies.

As much as this is a **_terrible practice_** and **_not a solution_**, I also went in and disable real-time AV.  Same error.  Also, there was nothing in the event logs to suggest anything found, and the control panel didn't show any history of blocking any operation.  Thus, there is no evidence to support AV being the cause.


* Is there more advanced logging available that would help deduce the cause?
* Is there a way for me to manually run the failing command?  (e.g., what directory should be the current working directory, if I attempt to manually execute the below command?)

```
C:/src/bug/vcpkg/downloads/tools/msys2/5f36c2d65b270362/usr/bin/bash.exe --noprofile --norc --debug -c "V=1 CPP='compile cl.exe -E' CC='compile cl.exe' CC_FOR_BUILD='compile cl.exe' CPP_FOR_BUILD='compile cl.exe -E' CXX_FOR_BUILD='compile cl.exe' CXX='compile cl.exe' RC='windres-rc rc.exe' WINDRES='windres-rc rc.exe' AR='ar-lib lib.exe' LD='link.exe -verbose' RANLIB=':' STRIP=':' NM='dumpbin.exe -symbols -headers' DLLTOOL='link.exe -verbose -dll' CCAS=':' AS=':' ./../src/libffi-3-b5bf0e8df1.clean/configure --build=x86_64-pc-mingw32 \"--enable-portable-binary\" \"--disable-docs\" \"--disable-multi-os-directory\" \"CFLAGS=${CFLAGS} -DFFI_BUILDING -DFFI_BUILDING_DLL\" \"CCAS=msvcc.sh -m64\" \"ac_cv_prog_ac_ct_STRIP=:\" \"gl_cv_double_slash_root=yes\" \"ac_cv_func_memmove=yes\" \"--disable-silent-rules\" \"--verbose\" \"--enable-shared\" \"--disable-static\" \"--prefix=/C/src/bug/vcpkg/installed/x64-windows/debug\" \"--bindir=\\${prefix}/../tools/libffi/debug/bin\" \"--sbindir=\\${prefix}/../tools/libffi/debug/sbin\" \"--libdir=\\${prefix}/lib\" \"--includedir=\\${prefix}/../include\" \"--datarootdir=\\${prefix}/share/libffi\""
```

**Comment by henrygab** [comment]
Running the command manually shows there's something wrong ... and it has nothing to do with AV:

```ps
PS C:\src\bug\vcpkg_installed\vcpkg\blds\libffi\x64-windows-dbg> C:/Users/lego/AppData/Local/vcpkg/downloads/tools/msys2/d7266db249278763/usr/bin/bash.exe --noprofile --norc --debug -c "V=1 CPP='compile cl.exe -E' CC='compile cl.exe' CC_FOR_BUILD='compile cl.exe' CPP_FOR_BUILD='compile cl.exe -E' CXX_FOR_BUILD='compile cl.exe' CXX='compile cl.exe' RC='windres-rc rc.exe' WINDRES='windres-rc rc.exe' AR='ar-lib lib.exe' LD='link.exe -verbose' RANLIB=':' STRIP=':' NM='dumpbin.exe -symbols -headers' DLLTOOL='link.exe -verbose -dll' CCAS=':' AS=':' ./../src/libffi-3-b5bf0e8df1.clean/configure --build=x86_64-pc-mingw32 \"--enable-portable-binary\" \"--disable-docs\" \"--disable-multi-os-directory\" \"CFLAGS=${CFLAGS} -DFFI_BUILDING -DFFI_BUILDING_DLL\" \"CCAS=msvcc.sh -m64\" \"ac_cv_prog_ac_ct_STRIP=:\" \"gl_cv_double_slash_root=yes\" \"ac_cv_func_memmove=yes\" \"--disable-silent-rules\" \"--verbose\" \"--enable-shared\" \"--disable-static\" \"--prefix=/C/src/bug/vcpkg_installed/x64-windows/debug\" \"--bindir=\\${prefix}/../tools/libffi/debug/bin\" \"--sbindir=\\${prefix}/../tools/libffi/debug/sbin\" \"--libdir=\\${prefix}/lib\" \"--includedir=\\${prefix}/../include\" \"--datarootdir=\\${prefix}/share/libffi\""
\--disable-docs\: -c: line 1: unexpected EOF while looking for matching `"'
```


@JonLiu1993 -- Did you run `vcpkg install` from within VS2022's command terminal (which is Powershell), or did you run it from a `cmd.exe` window?

**Comment by dg0yt** [comment]
The reported command line is not suitable for direct copy and paste.
Might be interesting to see you run `C:/src/bug/vcpkg/downloads/tools/msys2/023cdb3ca06f77f2/usr/bin/bash`, or maybe
~~~
C:/src/bug/vcpkg/downloads/tools/msys2/023cdb3ca06f77f2/usr/bin/bash -c env
~~~
which will print your environment incl. `PATH`.

**Comment by henrygab** [comment]
### Good news...
I think I found a cause, and want to see if we can make `vcpkg` more robust for its users, pointing folks to the real cause the issue.

When I started the msys bash, I immediately got errors about `fork()` failing.  Researching this more, it appears that `MSYS` is incompatible with basic security features:  Specifically, it [does not support mandatory ASLR](https://github.com/desktop/desktop/blob/development/docs/known-issues.md#enable-mandatory-aslr-triggers-cygheap-errors).

To test if this was the cause, I disabled mandatory ASLR system-wide.  `vcpkg install` succeeded, both on this minimal repro case and for my real-world scenario.

Conclusion:  **_Mandatory ASLR breaks MSYS_**.

Ok... let's see if an experienced user can realistically still have this security feature enabled by adding the appropriate exceptions.   Hmmm... the only types of exceptions allowed are either (a) giving an executable name... but that will match any path (_poor security choice_), or (b) listing the full path to a **_single_** executable.  Specifically, there is no ability to disable ASLR for all executables under a given tree.   If that had existed, then the exception would be covered by `%LOCALAPPDATA%/vcpkg/downloads/tools/msys2/...`, and/or `./vcpkg/downloads/tools/msys2/...`.

The limited options for exceptions cause two problems for users:
1. Not knowing which directory and subdirectory (hash/version) will be used (and updates will change which directory is used ...)
2. Even if knowning the directory, there are 28-300(!) individual executables (and thus potential exceptions to the ASLR settings).

Therefore, I will have two requests, which I will put into my next post.....

**Comment by henrygab** [comment]
Two requests:  The first is as an immediate improvement / bandaid.  The second is a longer-term, more comprehensive fix.

### Can `vcpkg` detect whether mandatory ASLR is set, and output a warning to the user about known incompatibilities with MSYS2?

While this won't solve the issue, it will allow users to determine the cause of otherwise seemingly-random failures.   Preferably, there would be a link to instructions on how to disable mandatory ASLR.  This may reduce time tracking down (or failing to root-cause) user issues, so I am hopeful this can be implemented rapidly ... as the cost/benefit is strongly in favor.

### Longer term, can `vcpkg` detect which directory it will run MSYS2 from, and add the corresponding exceptions to mandatory ASLR?

This would allow systems with all security features generally enabled to still use `vcpkg`.  Users would no longer need to choose between security features and functionality.   It would require a feature with test plan, but presumably have a shelf-life (when `MSYS2` fixes its incompatibility with ASLR).

Please let me know if either / both are possible?

**Comment by henrygab** [comment]
@JonLiu1993 ... are my requests simply out-of-scope, or are they initiating discussions about whether it's something reasonably possible to do, or ... something else?

**Comment by dg0yt** [comment]
@JonLiu1993 I assume the VS Team knows ASLR, so please forward it as needed.

I wonder if Mandatory ASLR adds much security to a machine which runs arbitrary build scripts. But vcpkg (at least `vcpkg_find_acquire_msys`) should offer better diagnostics. We can't expect msys/cygwin to generally enable ASLR soon, and I'm not aware of an alternative. Would https://learn.microsoft.com/en-us/powershell/module/processmitigations/get-processmitigation give the desired information?

**Comment by JonLiu1993** [comment]
@JavierMatosD, could you please take a look? thanks.

**Comment by BillyONeal** [comment]
I'm not sure how I feel about this; someone enabling EMET has very explicitly asked to be opted in to breakage like this; that's why the changes EMET makes are not defaults. vcpkg attempting to undo what EMET does would be wrong for the same reason we don't run `apt install` on users' behalf.

I would support grepping logs on failure for `cygheap base mismatch detected` and linking to that issue though.

**Comment by henrygab** [comment]
For avoidance of doubt, please review the [request](https://github.com/microsoft/vcpkg/issues/35966#issuecomment-1874490446):

> want to see if we can make vcpkg more robust for its users, pointing folks to the real cause the issue.

This is not EMET ... this is a feature that Windows 11 exposes directly to consumers.  Mandatory ASLR has a known compatibility issue with msys, whose implementation of `fork()` won't be changed (anytime soon) to support mandatory ASLR.  The methods of allowed exceptions (also detailed above) prevents manual exceptions being a reasonable user solution.

Thus, the question becomes (reworded from the above quote):

**_Can this be made less painful for the user?_**  Specifically, rather than an obscure failure message, can this condition be detected and notify the user of the problem?

Without that, the user is left thinking vcpkg is broken, but folks cannot repro the issue (even with 100% repro by the affected user).  In fact, you might discover this to have been a cause in other issues closed as "not repro".   Notification will not only improve perception of how awesome vcpkg is, it may save both you and the user time as the issue won't be filed (or resolved quickly).

**Comment by dg0yt** [comment]
> I would support grepping logs on failure for `cygheap base mismatch detected` and linking to that issue though.

If the reported console output is complete, then there are no logs to be grepped. However, there is error code 127:
~~~
-- Configuring x64-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: C:/src/bug/vcpkg/downloads/tools/msys2/5f36c2d65b270362/usr/bin/bash.exe --noprofile --norc --debug -c "V=1 CPP='compile cl.exe -E' CC='compile cl.exe' CC_FOR_BUILD='compile cl.exe' CPP_FOR_BUILD='compile cl.exe -E' CXX_FOR_BUILD='compile cl.exe' CXX='compile cl.exe' RC='windres-rc rc.exe' WINDRES='windres-rc rc.exe' AR='ar-lib lib.exe' LD='link.exe -verbose' RANLIB=':' STRIP=':' NM='dumpbin.exe -symbols -headers' DLLTOOL='link.exe -verbose -dll' CCAS=':' AS=':' ./../src/libffi-3-b5bf0e8df1.clean/configure --build=x86_64-pc-mingw32 \"--enable-portable-binary\" \"--disable-docs\" \"--disable-multi-os-directory\" \"CFLAGS=${CFLAGS} -DFFI_BUILDING -DFFI_BUILDING_DLL\" \"CCAS=msvcc.sh -m64\" \"ac_cv_prog_ac_ct_STRIP=:\" \"gl_cv_double_slash_root=yes\" \"ac_cv_func_memmove=yes\" \"--disable-silent-rules\" \"--verbose\" \"--enable-shared\" \"--disable-static\" \"--prefix=/C/src/bug/vcpkg/installed/x64-windows/debug\" \"--bindir=\\${prefix}/../tools/libffi/debug/bin\" \"--sbindir=\\${prefix}/../tools/libffi/debug/sbin\" \"--libdir=\\${prefix}/lib\" \"--includedir=\\${prefix}/../include\" \"--datarootdir=\\${prefix}/share/libffi\""
    Working Directory: C:/src/bug/vcpkg/buildtrees/libffi/x64-windows-dbg
    Error code: 127
    See logs for more information:

Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_make.cmake:832 (vcpkg_execute_required_process)
  ports/libffi/portfile.cmake:45 (vcpkg_configure_make)
  scripts/ports.cmake:170 (include)
~~~

**Comment by henrygab** [comment]
It's been ~45 days, so I thought I'd check if dg0yt's response enabled better error messaging?

My core question:
> Can this be made less painful for the user? Specifically, rather than an obscure failure message, can this condition be detected and notify the user of the actual problem?

**Comment by henrygab** [comment]
> @JavierMatosD, could you please take a look? thanks.

@JavierMatosD, were you able to review and respond as @JonLiu1993 asked?

It's now been months, so I thought I'd check if dg0yt's response enabled better error messaging?

My core question:
> Can VCPKG make this less painful for the user? Specifically, rather than an obscure failure message, can this condition be detected and then notify the user of the actual problem?

**Comment by BillyONeal** [comment]
Unfortunately exit code 127 doesn't really say anything in particular that I believe vcpkg could respond to. I don't think more time is going to change anything about this situation.

**Comment by henrygab** [comment]
The failure might not be translatable from error code 127.
However, the core issue could be detected (mandatory ASLR on a cygwin binary known to use `fork()`...).

Or, push have the Windows AppCompat database flags it as an incompatible configuration ... even if only to log a warning in the event log.   SOMETHING to give folks a fighting chance to understand what the real underlying issue is ... to avoid the long journeys down dead-ends as they try to figure this out....

**Comment by BillyONeal** [comment]
> However, the core issue could be detected (mandatory ASLR on a cygwin binary known to use `fork()`...).

In general it's fair to assume all cygwin binaries use that.

> Or, push have the Windows AppCompat database flags it as an incompatible configuration 

Windows already *does* have mitigations for stuff like that. Turing in EMET / Defender Exploit Guard is an explicit opt in saying "I prefer breakage to not having these mitigations turned on".

The user had to opt in to this situation. I don't think vcpkg trying to opt them back out again is correct. Depending on where the failure happens we might be able to issue a warning if we see 127 though.

**Comment by dg0yt** [comment]
> > However, the core issue could be detected (mandatory ASLR on a cygwin binary known to use `fork()`...).
> 
> In general it's fair to assume all cygwin binaries use that.

AFAIU the goal wasn't to detect the cygwin (here: MSYS2) binaries, but to detect the Windows ASLR setting, for a better diagnostic message.

**Comment by BillyONeal** [comment]
> AFAIU the goal wasn't to detect the cygwin (here: MSYS2) binaries, but to detect the Windows ASLR setting, for a better diagnostic message.

Different folks have been asking for different things in this issue over time; people want both a warning and us to somehow make it work. For a warning, I'm not sure responding to only the exit code is meaningful here, and I don't know how to practically detect this setting, so I haven't proposed anything specific.

For trying to "make it work", I feel that would be actively wrong. The only real way to make it work is to remove everything that comes from msys / cygwin as build time dependencies, but as far as I know there aren't practical alternatives for a lot of build systems. (After all, this is far from the only problem those dependencies have caused us)

**Comment by dg0yt** [comment]
> I don't know how to practically detect this setting, so I haven't proposed anything specific.

https://github.com/microsoft/vcpkg/issues/35966#issuecomment-1895388691 never received feedback IIRC

> to remove everything that comes from msys / cygwin as build time dependencies, but as far as I know there aren't practical alternatives for a lot of build systems.

Well, AFAICT the key hurdle is the `fork` facility, for ASLR and for some build time dependencies (in particular `sh`).

> After all, this is far from the only problem those dependencies have caused us.

I never saw plain cygwin being used here.

msys adds implicit path conversion heuristics to cygwin, and this comes with benefits and downsides. The problem of released binaries being removed from their servers is mitigated now by longer times of storage and by more pro-active updating in vcpkg.

I recently noticed https://github.com/skeeto/w64devkit offering another distribution of busybox GNU make, and a pkg-config implementation. But I didn't check if it could replace msys2 at least for some ports, and I don't know how it interacts with ASLR.

**Comment by BillyONeal** [comment]
> [#35966 (comment)](https://github.com/microsoft/vcpkg/issues/35966#issuecomment-1895388691) never received feedback IIRC

It's not obvious how that would affect anything here; it seems to only report things about a process that is already running? And it would add another powershell dependency and we've been trying to get rid of those.

> Well, AFAICT the key hurdle is the `fork` facility, for ASLR and for some build time dependencies (in particular `sh`).

Yes. That's why I said 'msys / cygwin' as both seem affected by this problem. That software that cares exists is the entire reason the switches to disable ASLR are available.

**Comment by henrygab** [comment]
> Well, AFAICT the key hurdle is the `fork` facility,

Yes, that's correct ... the underlying issue is `fork` being incompatible with ASLR, because the address space is not (cannot be) randomized.

> Different folks have been asking for different things in this issue over time; 

Once [the cause was determined](https://github.com/microsoft/vcpkg/issues/35966#issuecomment-1874490446), I posted [two asks](https://github.com/microsoft/vcpkg/issues/35966#issuecomment-1874495059):

1. Since vcpkg knows about this issue, can vcpkg **`detect`** whether mandatory ASLR is set, and **`output a warning`** to the user about known incompatibilities with MSYS2?

2. ~~Longer term, can vcpkg detect which directory it will run MSYS2 from, and add the corresponding exceptions to mandatory ASLR?~~

The second ask is clearly out-of-scope, so while I put it here for completeness, _**let's just focus on the first ask**_.  As the original person opening this issue, if others have different asks, they can open their own issue and link it here.  :)


### Detect and output a warning

Today, if a system has mandatory ASLR enabled (from whatever source, whether EMET or otherwise), the user is presented with non-actionable errors.   Worse, those errors do not naturally result in considering ASLR to be a cause.  As a result, the end user is left with misleading, cryptic error messages, and thus no signposts on how to understand the cause, or even correlate the failure with ASLR settings.

Without knowing the cause, a user cannot make an informed choice:  accept the cost(s) of mandatory ASLR, accept the costs(s) of disabling ASLR, etc.

As this is a known issue, having `vcpkg` output a warning message does at least two things:
1. Provides a signpost for users on what may have caused the known issue.
2. Provides a clear indicator that, when they inevitably ask for support, the output includes this same signpost, grealy reducing time-to-resolution.

**Comment by BillyONeal** [comment]
> 1. Since vcpkg knows about this issue, can vcpkg **`detect`** whether mandatory ASLR is set, and **`output a warning`** to the user about known incompatibilities with MSYS2?

Do we have a way to do that we can actually ship where the cure would not be worse than the disease? I think adding an additional powershell dependency / process launch would be worse than the disease given that the same kinds of environments that tend to deploy this also tend to deploy constrained execution policies for it e.g. #17890 which is one of the reasons we've been trying to remove all the hard powershell dependencies.

**Comment by henrygab** [comment]
I never said anything about powershell.  Are you saying that powershell is the only way to detect this is enabled?

**Comment by BillyONeal** [comment]
> I never said anything about powershell. Are you saying that powershell is the only way to detect this is enabled?

The only way that has been proposed thus far to detect this is PowerShell, yes.

**Comment by henrygab** [comment]
Didn't think that I would be the one to find and show this:

See https://msrc.microsoft.com/blog/2017/11/clarifying-the-behavior-of-mandatory-aslr/.

1. Read the `binary` registry value from `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\kernel\MitigationOptions`.
2. Read the low nibble of the second byte (e.g., `uint8_t v1 = uint8_array[1] & 0xFu`)
3. If that value is non-zero, then mandatory ASLR is enabled.

For the current purposes, that is sufficient to provide the warning to the user.
